File: output/figures/guilds/mal_guilds.py
#!/usr/bin/env python3
"""
figure/guilds/mal_guilds.py — build co-infection guild network from runs.tsv (MAL + HAL).

Treats all pathogens (primary + secondary) symmetrically. Nodes = pathogens,
edges = co-occurrence in the same confirmed run, edge weight = run count.

Normalises strain-level names to species (or f. sp.) level.
Kingdom is resolved via the phibase_db name_to_taxid lookup.

Outputs (figure/guilds/):
  mal_guild_nodes.tsv    name, kingdom, n_primary, n_secondary, total
  mal_guild_edges.tsv    node1, node2, weight

Run from crypt/:
  python figure/guilds/mal_guilds.py
"""

import csv
import json
import re
from collections import defaultdict
from pathlib import Path

RUNS_TSV        = Path("output/02_filter_runs/data/runs.tsv")
DB_PATH         = Path("output/00_build/data/phibase_db.json")
NODES_TSV       = Path("output/figures/guilds/mal_guild_nodes.tsv")
EDGES_TSV       = Path("output/figures/guilds/mal_guild_edges.tsv")

MIN_EDGE_WEIGHT = 2    # drop edges seen in fewer than this many runs
_SKIP           = {"environmental samples"}


# ── Name normalisation ────────────────────────────────────────────────────────

_FSP_RE    = re.compile(r'^(\S+\s+\S+\s+f\.\s+sp\.\s+\S+)')
_PCT_RE    = re.compile(r':[\d.]+%$')


def _normalize(name: str) -> str:
    """Collapse strain identifiers; keep virus common names and f. sp. epithets intact."""
    name = _PCT_RE.sub('', name).strip()
    # Virus/viroid common names are multi-word; don't truncate them
    if re.search(r'vir(?:us|oid)', name, re.IGNORECASE):
        return name
    # Keep f. sp. designation
    m = _FSP_RE.match(name)
    if m:
        return m.group(1)
    # Truncate strain-appended binomials to genus + species epithet only
    parts = name.split()
    return " ".join(parts[:2]) if len(parts) >= 2 else name


# ── DB kingdom lookup ─────────────────────────────────────────────────────────

def _build_kingdom_map(db: dict) -> tuple[dict[int, str], dict[str, int]]:
    """Return (taxid→kingdom, lowercase_name→taxid) from phibase_db."""
    taxid_to_kg: dict[int, str] = {}
    for key, kg in (("fungal_to_seed",    "Fungi"),
                    ("bacterial_to_seed",  "Bacteria"),
                    ("oomycete_to_seed",   "Oomycota"),
                    ("nematode_to_seed",   "Nematoda"),
                    ("virus_to_seed",      "Viruses")):
        for tid in db.get(key, {}):
            taxid_to_kg[int(tid)] = kg
    name_to_tid = {k: int(v) for k, v in db.get("name_to_taxid", {}).items()}
    return taxid_to_kg, name_to_tid


def _kingdom(name: str, taxid_str: str,
             taxid_to_kg: dict, name_to_tid: dict) -> str:
    """Resolve kingdom for a pathogen name/taxid pair."""
    if taxid_str:
        try:
            kg = taxid_to_kg.get(int(taxid_str))
            if kg:
                return kg
        except ValueError:
            pass
    tid = name_to_tid.get(name.lower()) or name_to_tid.get(_normalize(name).lower())
    return taxid_to_kg.get(tid, "Unknown") if tid else "Unknown"


# ── Main ──────────────────────────────────────────────────────────────────────

def main() -> None:
    with open(DB_PATH) as f:
        db = json.load(f)
    taxid_to_kg, name_to_tid = _build_kingdom_map(db)

    node_kingdom:     dict[str, str] = {}
    node_n_primary:   dict[str, int] = defaultdict(int)
    node_n_secondary: dict[str, int] = defaultdict(int)
    edge_counts:      dict[tuple[str, str], int] = defaultdict(int)

    with open(RUNS_TSV) as f:
        for row in csv.DictReader(f, delimiter='\t'):
            if not row.get('stat_pathogens'):
                continue

            # Primary node
            prim_norm = _normalize(row['library_organism'])
            prim_kg   = _kingdom(row['library_organism'], '',
                                 taxid_to_kg, name_to_tid)
            node_kingdom.setdefault(prim_norm, prim_kg)
            node_n_primary[prim_norm] += 1

            # Secondary nodes
            sec_norms: list[str] = []
            for entry in row['stat_pathogens'].split('; '):
                raw = _PCT_RE.sub('', entry).strip()
                if raw.lower() in _SKIP or not raw:
                    continue
                norm = _normalize(raw)
                kg   = _kingdom(raw, '', taxid_to_kg, name_to_tid)
                node_kingdom.setdefault(norm, kg)
                node_n_secondary[norm] += 1
                sec_norms.append(norm)

            # Co-occurrence pairs for this run
            run_pathogens = sorted({prim_norm} | set(sec_norms))
            for i, a in enumerate(run_pathogens):
                for b in run_pathogens[i + 1:]:
                    edge_counts[(a, b)] += 1

    # Write nodes
    with open(NODES_TSV, 'w', newline='') as f:
        w = csv.writer(f, delimiter='\t')
        w.writerow(['name', 'kingdom', 'n_primary', 'n_secondary', 'total'])
        for name in sorted(node_kingdom):
            n_p = node_n_primary.get(name, 0)
            n_s = node_n_secondary.get(name, 0)
            w.writerow([name, node_kingdom[name], n_p, n_s, n_p + n_s])

    # Write edges (filtered)
    with open(EDGES_TSV, 'w', newline='') as f:
        w = csv.writer(f, delimiter='\t')
        w.writerow(['node1', 'node2', 'weight'])
        for (a, b), wt in sorted(edge_counts.items(), key=lambda x: -x[1]):
            if wt >= MIN_EDGE_WEIGHT:
                w.writerow([a, b, wt])

    n_edges = sum(1 for wt in edge_counts.values() if wt >= MIN_EDGE_WEIGHT)
    print(f"Nodes: {len(node_kingdom)}")
    print(f"Edges (weight >= {MIN_EDGE_WEIGHT}): {n_edges}")
    print(f"Written: {NODES_TSV}, {EDGES_TSV}")


if __name__ == '__main__':
    main()

File: output/figures/guilds/mal_guilds.R
#!/usr/bin/env Rscript
# figure/guilds/mal_guilds.R — co-infection guild network (MAL + HAL)
# Run from crypt/: Rscript figure/guilds/mal_guilds.R

library(igraph)
library(ggraph)
library(ggplot2)
library(dplyr)

NODES_TSV <- "figure/guilds/mal_guild_nodes.tsv"
EDGES_TSV <- "figure/guilds/mal_guild_edges.tsv"
OUT_PDF   <- "figure/guilds/mal_guild_network.pdf"
OUT_PNG   <- "figure/guilds/mal_guild_network.png"

MIN_NODE_TOTAL <- 2    # drop nodes never appearing in a kept edge
LABEL_THRESHOLD <- 8   # label nodes with total detections >= this

# ── Colours ───────────────────────────────────────────────────────────────────
KINGDOM_COLS <- c(
  Fungi     = "#d35400",
  Viruses   = "#c0392b",
  Bacteria  = "#2980b9",
  Oomycota  = "#8e44ad",
  Nematoda  = "#27ae60",
  Unknown   = "#95a5a6"
)

# ── Load data ──────────────────────────────────────────────────────────────────
nodes <- read.delim(NODES_TSV, stringsAsFactors = FALSE)
edges <- read.delim(EDGES_TSV, stringsAsFactors = FALSE)

# Keep only nodes that appear in at least one edge
nodes_in_edges <- unique(c(edges$node1, edges$node2))
nodes <- nodes %>% filter(name %in% nodes_in_edges)

# ── Build igraph ───────────────────────────────────────────────────────────────
g <- graph_from_data_frame(
  d        = edges,
  vertices = nodes,
  directed = FALSE
)

# Node attributes
V(g)$kingdom <- nodes$kingdom[match(V(g)$name, nodes$name)]
V(g)$total   <- nodes$total[match(V(g)$name, nodes$name)]
V(g)$label   <- ifelse(nodes$total[match(V(g)$name, nodes$name)] >= LABEL_THRESHOLD,
                       V(g)$name, NA_character_)

# Shorten long virus names for label readability
V(g)$label <- sub("\\s+virus$", " virus", V(g)$label)   # no-op but keeps intent clear
V(g)$label <- gsub("f\\. sp\\. ", "f.sp.", V(g)$label)  # compact f. sp.

# ── Layout ─────────────────────────────────────────────────────────────────────
set.seed(42)
layout <- create_layout(g, layout = "fr")

# ── Plot ───────────────────────────────────────────────────────────────────────
p <- ggraph(layout) +
  geom_edge_link(
    aes(width = weight, alpha = weight),
    colour = "#666666",
    show.legend = FALSE
  ) +
  scale_edge_width(range = c(0.3, 3)) +
  scale_edge_alpha(range = c(0.2, 0.7)) +
  geom_node_point(
    aes(size = total, colour = kingdom),
    alpha = 0.9
  ) +
  geom_node_label(
    aes(label = label),
    size          = 2.2,
    label.padding = unit(0.12, "lines"),
    fill          = "white",
    colour        = "black",
    alpha         = 0.85,
    na.rm         = TRUE,
    repel         = FALSE
  ) +
  scale_colour_manual(values = KINGDOM_COLS, name = "Kingdom") +
  scale_size_continuous(
    name   = "Total detections",
    range  = c(2, 10),
    breaks = c(5, 20, 50, 100, 200)
  ) +
  labs(
    title    = "Cryptic co-infection guilds (MAL + HAL)",
    subtitle = "Nodes = pathogens; edges = co-occurrence in confirmed runs; edge width proportional to shared-run count"
  ) +
  theme_graph(base_family = "sans") +
  theme(
    legend.position  = "right",
    plot.title       = element_text(size = 12, face = "bold"),
    plot.subtitle    = element_text(size = 8, colour = "grey40"),
    legend.text      = element_text(size = 8),
    legend.title     = element_text(size = 9, face = "bold")
  )

ggsave(OUT_PDF, p, width = 12, height = 9)
ggsave(OUT_PNG, p, width = 12, height = 9, dpi = 200)
cat(sprintf("Written: %s, %s\n", OUT_PDF, OUT_PNG))
cat(sprintf("Nodes in graph: %d  |  Edges: %d\n", vcount(g), ecount(g)))
